### ISSUE TITLE
Feature request: Pipe input to docker build

### ISSUE THREAD

**Issue by MilesCranmer** [moby/moby#33556]
**Description**

*tl;dr - Is there any way I can pipe input to the* `docker build` *command?*

(I apologize if this functionality already exists, for I cannot find it)

My example is a bit trivial, but essentially whenever I am developing a library which has a dockerfile, I like to remove the installation of the library itself from the dockerfile, leaving just the dependencies. Then I append my "development environment" code to the bottom of the dockerfile, rebuild it, and then link the library directory so I can rapidly edit, build, and test it within the container.

In stead of manually adjusting this each time, I think it would be nice to have a single command that could set up my development container. It seems like it would be a small modification to the `docker build` command (if this isn't already feasible) to add this.

**Steps to ~~reproduce the issue~~ demonstrate the feature**

Essentially, what I am trying to do:

``` bash
$ ls
src
test
Dockerfile
README
$ docker build --stream $(append_dev_environment.sh Dockerfile 5) .
... #(builds)
```

In my case, `append_dev_environment.sh` would be:
``` bash
#!/bin/bash
# print out the current Dockerfile,
# removing the installation steps (last 5 lines here)
cat $1 | head -n -$2

# prints out my standalone development docker from github, 
# without the FROM statement (I gamble they will have similar base images)
wget -q .../Dockerfile -O - | tail -n +2 
```

As I mentioned above, this example is trivial and I can simply save a new dockerfile and build with the -f flag. But I could see this becoming useful as an automation tool for more complex Dockerfiles, instead of having to dump to a new file and build off of that.

**Comment by thaJeztah** [comment]
This looks like a duplicate of https://github.com/moby/moby/issues/27393, which is implemented in https://github.com/moby/moby/pull/31236, and part of docker 17.05 release; a simple example;

```bash
$ mkdir -p example && cd example
$ echo 'Hello!' > hello-world.txt
```

You can now use a Dockerfile passed as `stdin`, while still using the local build context (this was previously not possible);

```
$ docker build -t foo . -f-<<EOF
FROM alpine
COPY . .
CMD cat /hello-world.txt
EOF
```

After the image was built, you can see that the build-context is used from the local directory, and the `Dockerfile` that was taken from `stdin`;

```bash
$ docker run --rm foo
Hello!
```

Does that solve your issue?

**Comment by MilesCranmer** [comment]
Thanks, @thaJeztah. That looks like what I need. It appears my system's docker is a bit old, but I am excited to use the feature in future versions.
Cheers,
Miles

**Comment by thaJeztah** [comment]
Perfect!
